gatsby-node.js

```diff
@@ -189,7 +189,7 @@ exports.onCreateNode = ({node, boundActionCreators, getNode}) => {
           const day = match[3];
           const filename = match[4];
 
-          slug = `/blog/${year}/${month}/${day}/${filename}.html`;
+          slug = `blog/${year}/${month}/${day}/${filename}.html`;
```

Comment:
If we don't need slash here why do we need it on line 206 below?

Reply:
I've also notice that most permalink does not contain any forward slash at the beginning of the URL. I could not have find any possible way that any of the pages that falls into line 206. With the blog having the forward slash in front previously had causes an extra forward slash prepended. In line 206, I would not risk saying anything about it but if so, there could also be an extra slash prepended if it goes to line 206, based on the redirect of the `blog`.

<img width="741" alt="screen shot 2017-10-23 at 12 35 55 am" src="https://user-images.githubusercontent.com/22838512/31862483-38ee1886-b78a-11e7-970f-3aa35133c90d.png">

Reply:
@gaearon sorry, my bad. For line 206, that only applies to `index.md` in the root content folder. If somehow no _permalink_ is being supplied in any of the `md` file content other than `index.md` in the root content folder, it will give an error based on the slug being created. And because `index.md` does not have any parent to its _relativePath_ and it is a parent of itself, it is necessary for having the forward slash prepended in line 206.

Reply:
Maybe we can have something like below for line 206 to prevent unnecessary forward slash to be added if it's not the root `index.md` or coming from having its parent node like `blog`, `docs` for the slug creation. Given if there's a forward slash in the beginning and it does not come from root `index.md`, it will give error...

```js
if (relativePath.match(/.*\/index.md$/)) {
  // create slug with no forward slash in the beginning
} else {
  // create slug with forward slash in the beginning
  slug = `/${relativePath.replace('.md', '.html')}`;
}
```

Just a thought and findings that I made. Happy to hear what you have in mind about this PR 😃 @gaearon

Reply:
I just built the site locally (`yarn build`) and looked at the generated `public/_redirects` file (the file that's [used by Netlify for redirects](https://www.netlify.com/docs/redirects/)). Looking for this specific redirect I see:
```
/blog/2014/10/14/introducting-react-elements.html  //blog/2014/10/14/introducing-react-elements.html  302
```

The double slash before "blog" shouldn't be there. That problem only occurs for this specific redirect. I _think_ that's because our redirect-creation loop always prepends a "/" before the `slug`:
```js
          createRedirect({
            fromPath: `/${fromPath}`,
            redirectInBrowser: true,
            toPath: `/${slug}`,
          });
```

I think the fix to this might just be to conditionally prefix:
```diff
diff --git a/gatsby-node.js b/gatsby-node.js
index 9613217f8..6846788f5 100644
--- a/gatsby-node.js
+++ b/gatsby-node.js
@@ -114,11 +114,13 @@ exports.createPages = async ({graphql, boundActionCreators}) => {
             process.exit(1);
           }
 
+          const toPath = slug.startsWith('/') ? slug : `/${slug}`
+
           redirectToSlugMap[fromPath] = slug;
           createRedirect({
             fromPath: `/${fromPath}`,
             redirectInBrowser: true,
-            toPath: `/${slug}`,
+            toPath,
           });
         });
       }
```

That change corrects the entry in the `public/_redirects` file _and_ fixes the (local) redirect logic in the dev server.